MapboxCoreNavigation/CLLocation.swift

```diff
@@ -147,11 +148,15 @@ extension CLLocation {
     /**
      Determines if the a location is qualified enough to allow the user puck to become unsnapped.
      */
-    func shouldSnap(toRouteWith course: CLLocationDirection) -> Bool {
+    func shouldSnap(toRouteWith course: CLLocationDirection, distanceToFirstCoordinateOnLeg: CLLocationDistance = CLLocationDistanceMax) -> Bool {
+        
+        // If the user is near the beginning of leg, allow for unsnapped more often.
+        let isWithinDepatureStep = distanceToFirstCoordinateOnLeg < RouteControllerManeuverZoneRadius
+
         if course >= 0 &&
-            speed >= RouteSnappingMinimumSpeed &&
-            course.differenceBetween(self.course) > RouteSnappingMaxManipulatedCourseAngle &&
-            horizontalAccuracy < RouteSnappingMinimumHorizontalAccuracy {
+            (speed >= RouteSnappingMinimumSpeed || isWithinDepatureStep) &&
+            (horizontalAccuracy < RouteSnappingMinimumHorizontalAccuracy || isWithinDepatureStep) &&
```

Comment:
@bsudekum hmmm, we check for `isWithinDepatureStep` twice in both conditions.

Reply:
@vincethecoder what would you suggest we do otherwise here?

Reply:
@bsudekum this could be refactored as:
```
(speed >= RouteSnappingMinimumSpeed || isWithinDepatureStep || horizontalAccuracy < RouteSnappingMinimumHorizontalAccuracy) && ...```

Reply:
I agree that works, but I think breaking it out line by line makes it more clear what is happening here.

Reply:
We can eliminate the `isWithinDepatureStep` check in one of the 2 conditions.

Reply:
@vincethecoder Right I get that, but I think by breaking it out, line by line, it's easy for someone to glance at what is happening here.

Reply:
@bsudekum hmmm the concern is `isWithinDepatureStep` checked in both conditions might be covering a defect that we need to address. This is what I mean:
```
(speed >= RouteSnappingMinimumSpeed [evaluates false] || isWithinDepatureStep [evaluates true]) && (horizontalAccuracy < RouteSnappingMinimumHorizontalAccuracy [evaluates false] || isWithinDepatureStep [evaluates true])
```
In this case, we are forcing the 2 conditions to evaluate to `true`, should `isWithinDepatureStep evaluate as `true` (regardless of the left conditions). This is a bit of a concern. Do you see what I mean?

**Proposed fix for readability**:
```
(speed >= RouteSnappingMinimumSpeed || isWithinDepatureStep) &&
(horizontalAccuracy < RouteSnappingMinimumHorizontalAccuracy) && ...
```
this will be less risky and still readable.

Reply:
@vincethecoder the proposed code will not work. If `(horizontalAccuracy < RouteSnappingMinimumHorizontalAccuracy` evaluates to false, we should not snap if `isWithinDepatureStep`  is true.

I made simple example you can paste into a playground here: https://gist.github.com/bsudekum/bfcab90b6ed9f04ffd141850251235e4

Reply:
I think we are splitting hairs here though. I think the proposed code is fine as is.